### ISSUE TITLE
Zip Streams

### ISSUE THREAD

**Issue by nathanila** [amaembo/streamex#60]
Please add a method of the signature: EntryStream<K, V> EntryStream.zip(Stream<K> keys, Stream<V> values).  This will allow for zipping two streams together.  See this Stack Overflow question for more details.

http://stackoverflow.com/questions/17640754/zipping-streams-using-jdk8-with-lambda-java-util-stream-streams-zip

**Comment by amaembo** [comment]
Thank you for proposal. I already thought about such feature and unfortunately again I don't like it. There are several reasons (probably partially the same which lead to removal of `zip` from JDK). First, zipped stream cannot be efficiently parallelized even if both original streams can. That's because original streams can split at different positions and they even don't have to report the splitting position. Second, even for non-short-circuiting operation you cannot use `Spliterator.forEachRemaining` internally. For some streams `Spliterator.tryAdvance` is much less efficient. Third, having parallel streams (where n-th element of the first stream corresponds to the n-th element of the second stream) actually smells. This might be reasonable in purely functional languages where tuples rule. But Java is still object-oriented language. If you have `Stream<A>` and `Stream<B>` and first A corresponds to the first B, how did you get such streams? Why don't you have `Stream<AB>` where `AB` is the object which encapsulates both A and B?

Finally we already have `EntryStream.zip(List, List)` and `EntryStream.zip(A[], B[])` which are free of the first two problems. So new method would be useful for stream sources which are not arrays or random access lists. Also if you want to zip the list with indices (like `zip(IntStream.range(0, list.size()).boxed(), list.stream())`), then we already have `EntryStream.of(List)`.

If you have some other practical use-cases when zipped stream is necessary and cannot be expressed with existing methods, please tell me. In general I tend to think that adding `zip(Stream, Stream)` may cause the overuse of such method instead of more efficient alternatives.

**Comment by nathanila** [comment]
Because StreamEx is powerful, I am switching over to returning streams from methods and passing streams as arguments instead of using List.  In several situations, this avoids having to convert a stream into a List, return the List and then only to have the caller create a stream from the List.

My use case has a stream passed as an argument to a method.  The stream is going to be put into a Properties object which is then going to be stored on disk.  The keys are a combination of a prefix and the index.

   private static final String RECENT_PREFIX = "recent.";
   private static final int RECENT_LIMIT = 10;

   private static void setRecent(Properties properties, StreamEx<File> recent)
   {
      StreamEx<String> keys;
      long count;

      keys = IntStreamEx.
         range(0, RECENT_LIMIT).
         mapToObj(index -> RECENT_PREFIX + index);

      count = StreamUtils.
         zip(keys, recent).                                         // Creates an EntryStream<String, File>
         mapValues(File::getAbsolutePath).
         peekKeyValue(properties::setProperty).
         count();

      IntStreamEx.
         range((int) count, RECENT_LIMIT).
         mapToObj(index -> RECENT_PREFIX + index).
         forEach(properties::remove);
   }

As you suggested, I could create a KeyFile object.  However, I am not sure how to create the keys while going through the recent stream.

**Comment by amaembo** [comment]
While returning stream from the method is perfectly fine, passing it to the method is sometimes questionable as this limits the method implementation to use the passed stream only once. In this particular case passing the `List` looks more natural to me. Anyways having `RECENT_LIMIT = 10` it's not a big deal to dump it to the list:

    List<File> list = recent.limit(RECENT_LIMIT).toList();
    int count = list.size();
    EntryStream.of(list)
               .mapKeys(index -> RECENT_PREFIX + index)
               .mapValues(File::getAbsolutePath)
               .forKeyValue(properties::setProperty);

   IntStreamEx.
      range(count, RECENT_LIMIT).
      mapToObj(index -> RECENT_PREFIX + index).
      forEach(properties::remove);

In general do not try to do everything in the single pipeline. Sometimes it makes code harder to maintain and adds no value in terms of performance.

Ok, I'll rethink the `zip` feature, though I don't promise anything. Probably it should be `StreamEx.zipWith` instance method, so you can write `keys.zipWith(recent).mapValues(...)`. There are many open questions like whether the resulting stream should be parallel, what should we do if streams have different length, can we perform some optimizations for SIZED/SUBSIZED sources and so on.

By the way your code is dangerous. Using `peek` with terminal `count` may fail in JDK-9 where `count` for sized stream does not traverse the input at all.

**Comment by nathanila** [comment]
Thank you for the code suggestion.

 

> whether the resulting stream should be parallel

 

When I was writing my implementation, I struggled with that too.  I think that depends if the source streams are parallel and if the implementation can be executed in parallel.

 

I am not sure how you are going to be able to have 2 threads pulling from 2 streams without getting the elements mixed in the wrong order.  For example, lets say Thread A pulls from the first stream.  Then, Thread B pulls from the first stream and the second stream.  Then, Thread A pulls from the second stream.  Now, both threads have mismatched elements.  Thread A should have pulled from the second stream before Thread B pulls from the second stream.

 

Splitting streams won’t work because we can’t force both split streams to have the same number of elements.

It turns out that I took the easy route and not necessarily the performant route.  I changed the streams to Iterators and then implemented an Iterator which pulls from both.  I then convert the Iterator back into a stream.  See this answer.  http://stackoverflow.com/questions/17640754/zipping-streams-using-jdk8-with-lambda-java-util-stream-streams-zip/32342172#32342172 

 

> what should we do if streams have different length

 

I see that the existing StreamEx.zip() methods throw an exception.  In my use case, that would be very tricky to get the stream lengths to match.  

 

Consider simply stopping when either stream comes to the end.  Consider implementing zip(Stream<A> first, Stream<B> second) but also implement zip(Stream<A> first, Stream<B> second, Consumer<A> extraFirst, Consumer<B> extraSecond).  In the latter case, the extra elements are dumped into the consumers.

 

> can we perform some optimizations for SIZED/SUBSIZED sources

 

You have a lot of good questions.  I am glad you are developing StreamEx and not me.  I would just plough ahead and miss a lot of important cases.  Please give zip stream a thought and let me know what you decide.  Your insights are very helpful.

 

> Using peek with terminal count may fail in JDK-9 where count for sized stream does not traverse the input at all.

 

Scary!  That’s a great optimization but only if the intermediate operators allow for it.  I would think peek and filter would not allow for it.  Yet, map, skip and limit would allow for it.

 

-Nathan

**Comment by amaembo** [comment]
The simple parallelization optimization would be to trySplit() both original streams if both are SUBSIZED and their size is exactly the same. If they *happens* to split at the same position, lucky-lucky, we can nicely split zipped stream as well. Otherwise we concatenate splitted parts back (I already have my own `TailConcatSpliterator`, see #55, so no need to meddle with `Stream.concat`) and fall-back to poor-man implementation (dumping parts of the stream into array). This optimization would work if both streams are created from `ArrayList` or array and perform only `map/boxed/etc.` operations upstream. Sometimes it would work with `IntStream.range` stream (depending on its exact size).

If streams split unevenly, but split-points are not too far from each other, it's possible to implement some sophisticated algorithm involving buffering (the buffer should be shared between two spliterators working in different threads). It's hard, but possible. Probably can be postponed.

If both input streams are SIZED, but not SUBSIZED, we can still use the size for more balanced poor-man splitting. This would still be better than `spliteratorUnknownSize` (which is the worst way to split actually).

**Comment by amaembo** [comment]
Draft implementation without splitting is committed. I decided to ignore the longer stream tail. This seems to be quite expected behavior. At least you can zip the stream with indices using `myStreamEx.zipWith(IntStreamEx.range(Integer.MAX_VALUE).boxed())` without knowing the exact size of initial stream. Well actually probably it would be good to implement separate `EntryStream<Integer, T> withIndex()` method with separate spliterator. It can nicely split for SIZED/SUBSIZED input unlike `myStreamEx.zipWith(IntStreamEx.range(Integer.MAX_VALUE).boxed())`.

I still optimize `forEachRemaining` if both inputs are SIZED: we reuse `forEachRemaining` of the shorter stream. And, of course, I report the resulting stream size if both inputs are SIZED.

>  I would think peek and filter would not allow for it. Yet, map, skip and limit would allow for it.

`peek` does not clear SIZED characteristic, but `skip` and `limit` clear it in current implementation (though this might change in future).

Also by the way beware of this technique:

    final Iterable<T> iterable = () -> iterator;
    return StreamSupport.stream(iterable.spliterator(), parallel);

It creates unordered stream!

**Comment by nathanila** [comment]
Thank you for implementing this.

> beware of this technique:

Before I encountered StreamEx, I would do exactly that to create a stream from an iterator when I couldn't figure out another way to create a stream.  Now, that I have the power of StreamEx, I use `StreamEx.of(iterator)`.

I am truly grateful for StreamEx.  It provides the power and flexibility to really make streams more applicable in code.  Recently, I had to modify some code I wrote a while ago.  The original implementation used a for loop because the operations in the body couldn't be put into a Java 8 `Stream`.  With StreamEx, I was able re-implemented the for loop as a StreamEx!  This greatly simplified the code!
